libraries/standard/http/src/http_client.c

```diff
@@ -40,21 +139,149 @@ HTTPStatus_t HTTPClient_InitializeRequestHeaders( HTTPRequestHeaders_t * pReques
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( fieldLen == 0u )
+    {
+        IotLogError( "Parameter check failed: fieldLen must be greater than 0." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( valueLen == 0u )
+    {
+        IotLogError( "Parameter check failed: valueLen must be greater than 0." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else
+    {
+        /* Empty else MISRA 15.7 */
+    }
+
+    if( returnStatus == HTTP_SUCCESS )
+    {
+        returnStatus = _addHeader( pRequestHeaders,
+                                   pField, fieldLen, pValue, valueLen );
+    }
+
+    return returnStatus;
 }
 
 /*-----------------------------------------------------------*/
 
 HTTPStatus_t HTTPClient_AddRangeHeader( HTTPRequestHeaders_t * pRequestHeaders,
-                                        int32_t rangeStart,
+                                        int32_t rangeStartOrlastNbytes,
                                         int32_t rangeEnd )
 {
-    return HTTP_NOT_SUPPORTED;
+    HTTPStatus_t returnStatus = HTTP_SUCCESS;
+    char rangeValueBuffer[ MAX_RANGE_REQUEST_VALUE_LEN ] = { 0 };
+    size_t rangeValueLength = 0;
+    int stdRetVal = 0;
+
+    if( pRequestHeaders == NULL )
+    {
+        IotLogError( "Parameter check failed: pRequestHeaders is NULL." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( pRequestHeaders->pBuffer == NULL )
+    {
+        IotLogError( "Parameter check failed: pRequestHeaders->pBuffer is NULL." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( rangeEnd < HTTP_RANGE_REQUEST_END_OF_FILE )
+    {
+        IotLogErrorWithArgs( "Parameter check failed: rangeEnd is invalid: "
+                             "rangeEnd should be >=-1: RangeEnd=%d", rangeEnd );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( ( rangeStartOrlastNbytes < 0 ) &&
+             ( rangeEnd != HTTP_RANGE_REQUEST_END_OF_FILE ) )
+    {
+        IotLogErrorWithArgs( "Parameter check failed: Invalid range values: "
+                             "rangeEnd should be -1 when rangeStart < 0: "
+                             "RangeStart=%d, RangeEnd=%d",
+                             rangeStartOrlastNbytes, rangeEnd );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else if( ( rangeEnd != HTTP_RANGE_REQUEST_END_OF_FILE ) &&
+             ( rangeStartOrlastNbytes > rangeEnd ) )
+    {
+        IotLogErrorWithArgs( "Parameter check failed: Invalid range values: "
+                             "rangeStart should be < rangeEnd when both are >= 0: "
+                             "RangeStart=%d, RangeEnd=%d",
+                             rangeStartOrlastNbytes, rangeEnd );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+    else
+    {
+        /* Populate the buffer with the value data for the Range Request.*/
+        stdRetVal = sprintf( rangeValueBuffer,
```

Comment:
You must use "snprintf" not sprintf.

Reply:
It is not possible to estimate the size of the ASCII version of the range value, due to which I have used `sprintf` here

Reply:
snprintf simply takes the maximum size that can be written, we can write less.

Reply:
We really need to snprintf appsec will flag this and cut a ticket. And then we will have to waste time explaining away the flag.
```
snprintf( rangeValueBuffer,
              MAX_RANGE_REQUEST_VALUE_LEN + 1,
              "%s%d"
              RANGE_REQUEST_HEADER_VALUE_PREFIX,
              rangeStartOrlastNbytes );
```